README.md
BFPaperButton
=============
[![CocoaPods](https://img.shields.io/cocoapods/v/BFPaperButton.svg?style=flat)](https://github.com/bfeher/BFPaperButton)

> A flat button inspired by Google Material Design's Paper theme.

![Animated Screenshot](https://raw.githubusercontent.com/bfeher/BFPaperButton/master/BFPaperButtonDemoGif2.gif "Animated Screenshot")


About
---------
### Now with smoother animations and more public properties for even easier customization!

_BFPaperButton_ is a subclass of UIButton that behaves much like the new paper buttons from Google's Material Design Labs.
All animation are asynchronous and are performed on sublayers.
BFPaperButtons work right away with pleasing default behaviors, however they can be easily customized! The corner radius, tap-circle color, background fade color, and tap-circle diameter are all readily customizable via public properties.

BFPaperButtons come in 2 flavors, Flat or Raised. 
Flat BFPaperButtons have no shadow and will remain flat through all animations. Flat buttons can be transparent, in which case the background will also fade a little when tapped.
Raised BFPaperButtons have a drop shadow that animates along with a tap, giving it the feeling of raising up with your touch. Raised BFPaperButtons do not look good with a clear background color since it will expose their shadow layer underneath.

By default, BFPaperButtons use "Smart Color" which will match the tap-circle and background fade colors to the color of the `titleLabel`.
You can turn off Smart Color by setting the property, `.usesSmartColor` to `NO`. If you disable Smart Color, a gray color will be used by default for both the tap-circle and the background color fade.
You can set your own colors via: `.tapCircleColor` and `.backgroundFadeColor`. Note that setting these disables Smart Color.

***Note*** *that setting the button type to Custom in the Storyboard is required to prevent the title from fading out on UIControlStateHighlighted.*

## Properties

`UIColor *shadowColor`
> The UIColor for the shadow of a raised button. An alpha value of `1` is recommended as shadowOpacity overwrites the alpha of this color.

`CGFloat loweredShadowOpacity`
> A CGFLoat representing the opacity of the shadow of RAISED buttons when they are lowered (idle). Default is `0.5f`.

`CGFloat loweredShadowRadius`

> A CGFLoat representing the radius of the shadow of RAISED buttons when they are lowered (idle). Default is `1.5f`.

`CGSize loweredShadowOffset`

> A CGSize representing the offset of the shadow of RAISED buttons when they are lowered (idle). Default is `(0, 1)`.

`CGFloat liftedShadowOpacity`

> A CGFLoat representing the opacity of the shadow of RAISED buttons when they are lifted (on touch down). Default is `0.5f`.

`CGFloat liftedShadowRadius`

> A CGFLoat representing the radius of the shadow of RAISED buttons when they are lifted (on touch down). Default is `4.5f`.

`CGSize liftedShadowOffset`

> A CGSize representing the offset of the shadow of RAISED buttons when they are lifted (on touch down). Default is `(2, 4)`.

`CGFloat touchDownAnimationDuration`

> A CGFLoat representing the duration of the animations which take place on touch DOWN! Default is `0.25f` seconds. (Go Steelers)

`CGFloat touchUpAnimationDuration`

> A CGFLoat representing the duration of the animations which take place on touch UP! Default is `2 * touchDownAnimationDuration` seconds.

`CGFloat tapCircleDiameterStartValue`

> A CGFLoat representing the diameter of the tap-circle as soon as it spawns, before it grows. Default is `5.f`.

`CGFloat tapCircleDiameter`

> The CGFloat value representing the Diameter of the tap-circle. By default it will be the result of `MAX(self.frame.width, self.frame.height)`. `tapCircleDiameterFull` will calculate a circle that always fills the entire view. Any value less than or equal to `tapCircleDiameterFull` will result in default being used. The constants: `tapCircleDiameterLarge`, `tapCircleDiameterMedium`, and `tapCircleDiameterSmall` are also available for use. */

`CGFloat tapCircleBurstAmount`

> The CGFloat value representing how much we should increase the diameter of the tap-circle by when we burst it. Default is `100.f`.

`CGFloat cornerRadius`

> The corner radius which propagates through to the sub layers. Default is `0`.

`UIColor *tapCircleColor`

> The UIColor to use for the circle which appears where you tap. NOTE: Setting this defeats the "Smart Color" ability of the tap circle. Alpha values less than `1` are recommended.

`UIColor *backgroundFadeColor`
> The UIColor to fade clear backgrounds to. NOTE: Setting this defeats the "Smart Color" ability of the background fade. Alpha values less than `1` are recommended.

`BOOL rippleFromTapLocation`
> A flag to set to `YES` to have the tap-circle ripple from point of touch. If this is set to `NO`, the tap-circle will always ripple from the center of the view. Default is `YES`.

`BOOL rippleBeyondBounds`
> A flag to set to `YES` to have the tap-circle ripple beyond the bounds of the view. If this is set to `NO`, the tap-circle will be clipped to the view's bounds. Default is `NO`.

`BOOL isRaised`
> A flag to set to `YES` to CHANGE a flat view to raised, or set to `NO` to CHANGE a raised view to flat. If you used one of the provided custom initializers, you should probably leave this parameter alone. If you instantiated via storyboard or IB and want to CHANGE from riased to flat, this is the parameter for you! Default is `YES`.

`BOOL usesSmartColor`
> A flag to set YES to use Smart Color, or NO to use a custom color scheme. While Smart Color is the default (usesSmartColor = YES), customization is cool too.


### Notes on RAISED vs FLAT and SMART COLOR vs NON SMART COLOR:
##### RAISED
Has a shadow, so a clear background will look silly. It has only a tap-circle color. No background-fade.

##### FLAT
Has no shadow, therefore clear backgrounds look fine. If the background is clear, it also has a background-fade color to help visualize the button and its frame.

##### SMART COLOR
Will use the titleLabel's font color to pick a tap circle color and, if the background is clear, will also pick a lighter background fade color.

##### NON SMART COLOR
Will use a translucent gray tap-circle and, if the background is clear, a lighter translucent gray background-fade color.


Usage
---------
Add the _BFPaperButton_ header and implementation file to your project. (.h & .m)

### Creating a Flat BFPaperButton
```objective-c
BFPaperButton *flatPaperButton = [[BFPaperButton alloc] initWithFrame:rect raised:NO];
```

### Creating a Raised BFPaperButton
```objective-c
BFPaperButton *raisedPaperButton = [[BFPaperButton alloc] initWithFrame:rect raised:YES];
```

### Working Example
*(Taken directly from example project.)*<br />
```objective-c
BFPaperButton *bfFlatSmart = [[BFPaperButton alloc] initWithFrame:CGRectMake(20, 20, 280, 43) raised:NO];
[bfFlatSmart setTitle:@"BFPaperButton Flat: Smart Color" forState:UIControlStateNormal];
bfFlatSmart.backgroundColor = [UIColor paperColorGray600];	// You can find this from my other library, BFPaperColors :)
[bfFlatSmart setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
[bfFlatSmart setTitleColor:[UIColor whiteColor] forState:UIControlStateHighlighted];
[bfFlatSmart addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
[self.view addSubview:bfFlatSmart];
```

### Customized Example
*(Taken directly from example project.)*<br />
```objective-c
BFPaperButton *circle2 = [[BFPaperButton alloc] initWithFrame:CGRectMake(116, 468, 86, 86) raised:YES];
[circle2 setTitle:@"Custom" forState:UIControlStateNormal];
[circle2 setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
[circle2 setTitleColor:[UIColor whiteColor] forState:UIControlStateHighlighted];
[circle2 addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
circle2.backgroundColor = [UIColor colorWithRed:0.3 green:0 blue:1 alpha:1];
circle2.tapCircleColor = [UIColor colorWithRed:1 green:0 blue:1 alpha:0.6];  // Setting this color overrides "Smart Color".
circle2.cornerRadius = circle2.frame.size.width / 2;
circle2.rippleFromTapLocation = NO;
circle2.rippleBeyondBounds = YES;
circle2.tapCircleDiameter = MAX(circle2.frame.size.width, circle2.frame.size.height) * 1.3;
[self.view addSubview:circle2];
```

CocoaPods
-------

CocoaPods are the best way to manage library dependencies in Objective-C projects.
Learn more at http://cocoapods.org

Add this to your podfile to add BFPaperButton to your project.
```ruby
platform :ios, '7.0'
pod 'BFPaperButton'
```


License
--------
_BFPaperButton_ uses the MIT License:

> Please see included [LICENSE file](https://raw.githubusercontent.com/bfeher/BFPaperButton/master/LICENSE.md).


CHANGELOG.md
BFPaperButton
=============
[![CocoaPods](https://img.shields.io/cocoapods/v/BFPaperButton.svg?style=flat)](https://github.com/bfeher/BFPaperButton)

> Note that this changelog was started very late, at roughly the time between version 2.0.17 and 2.0.18. Non consecutive jumps in changelog mean that there were incremental builds that weren't released as a pod, typically while solving a problem.


2.1.2
---------
* (^) Removed warnings about CAAnimationDelegate in Xcode 8 (thanks [Nicola](https://github.com/ndebei) and [h-umus](https://github.com/h-umus)). [Pull Request 9](https://github.com/bfeher/BFPaperButton/pull/31)  


2.1.1
---------
* (^) Fixed bug where '.enabled' property was not being taken into account during sizeToFit and layoutSubviews calls. [Issue 28](https://github.com/bfeher/BFPaperButton/issues/28)  


2.1.0
---------
* (+) Added IB_DESIGNABLE/IBInspectable attributes to properties courtesy of github user [Bhargav Mogra](https://github.com/bhargavms)  
* (^) Updated README to be more readable courtesy of github user [Bhargav Mogra](https://github.com/bhargavms)  
* (-) Removed BFPaperColor dependency courtesy of github user [Bhargav Mogra](https://github.com/bhargavms)  


2.0.29
---------
* (+) Migrated to Cocoapods 1.0.


2.0.27
---------
* (+) Added LaunchScreen.xib to get fullscreen for iPhone 5 and up. (Lazy demo app doesn't scroll so iPhone 4 and down won't fit everything on screen.)  
* (^) Modified BFPaperButton.m to check if tapCircles are nil first (credit goes to github user [Adam Szeptycki](https://github.com/adamszeptycki))


2.0.18
---------
* (^) Fixed bug where loweredShadowOffset property wasn't being utilized and instead was ignored for a hardcoded default value.


BFPaperButton/BFPaperViewController.m
//
//  BFPaperViewController.m
//  BFPaperButton
//
//  Created by Bence Feher on 7/16/14.
//  Copyright (c) 2014 Bence Feher. All rights reserved.
//
// The MIT License (MIT)
//
// Copyright (c) 2014 Bence Feher
//
// Permission is hereby granted, free of charge, to any person obtaining a copy of
// this software and associated documentation files (the "Software"), to deal in
// the Software without restriction, including without limitation the rights to
// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
// the Software, and to permit persons to whom the Software is furnished to do so,
// subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


#import "BFPaperViewController.h"
#import "BFPaperButton.h"

@interface BFPaperViewController ()

@end

@implementation BFPaperViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view, typically from a nib.
    
    BFPaperButton *bfFlatSmart = [[BFPaperButton alloc] initWithFrame:CGRectMake(20, 20, 280, 43) raised:NO];
    [bfFlatSmart setTitle:@"BFPaperButton Flat: Smart Color" forState:UIControlStateNormal];
    [bfFlatSmart setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    bfFlatSmart.backgroundColor = [UIColor colorWithRed:117.f/255.f green:117.f/255.f blue:117.f/255.f alpha:1];
    [bfFlatSmart setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [bfFlatSmart setTitleColor:[UIColor whiteColor] forState:UIControlStateHighlighted];
    [bfFlatSmart addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:bfFlatSmart];
    
    BFPaperButton *bfFlatDumb = [[BFPaperButton alloc] initWithFrame:CGRectMake(20, 71, 280, 43) raised:NO];
    bfFlatDumb.usesSmartColor = NO;
    [bfFlatDumb setTitle:@"BFPaperButton Flat: !Smart Color" forState:UIControlStateNormal];
    [bfFlatDumb setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    bfFlatDumb.backgroundColor = [UIColor colorWithRed:117.f/255.f green:117.f/255.f blue:117.f/255.f alpha:1];
    [bfFlatDumb setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [bfFlatDumb setTitleColor:[UIColor whiteColor] forState:UIControlStateHighlighted];
    [bfFlatDumb addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:bfFlatDumb];
    
    BFPaperButton *bfFlatClearSmart = [[BFPaperButton alloc] initWithFrame:CGRectMake(20, 122, 280, 43) raised:NO];
    [bfFlatClearSmart setTitle:@"BFPaperButton Flat: Clear, Smart Color" forState:UIControlStateNormal];
    [bfFlatClearSmart setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    [bfFlatClearSmart setTitleColor:[UIColor colorWithRed:33.f/255.f green:150.f/255.f blue:243.f/255.f alpha:1] forState:UIControlStateNormal];
    [bfFlatClearSmart addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:bfFlatClearSmart];
    
    BFPaperButton *bfFlatClearDumb = [[BFPaperButton alloc] initWithFrame:CGRectMake(20, 173, 280, 43) raised:NO];
    bfFlatClearDumb.usesSmartColor = NO;
    [bfFlatClearDumb setTitle:@"BFPaperButton Flat: Clear, !Smart Color" forState:UIControlStateNormal];
    [bfFlatClearDumb setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    [bfFlatClearDumb setTitleColor:[UIColor colorWithRed:33.f/255.f green:150.f/255.f blue:243.f/255.f alpha:1] forState:UIControlStateNormal];
    [bfFlatClearDumb addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:bfFlatClearDumb];
    
    
    
    BFPaperButton *bfRaisedSmart = [[BFPaperButton alloc] initWithFrame:CGRectMake(20, 239, 280, 43) raised:YES];
    [bfRaisedSmart setBackgroundColor:[UIColor colorWithRed:33.f/255.f green:150.f/255.f blue:243.f/255.f alpha:1]];
    [bfRaisedSmart setTitle:@"BFPaperButton Raised: Smart Color" forState:UIControlStateNormal];
    [bfRaisedSmart setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    [bfRaisedSmart addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:bfRaisedSmart];
    
    BFPaperButton *bfRaisedDumb = [[BFPaperButton alloc] initWithFrame:CGRectMake(20, 307, 280, 43) raised:YES];
    bfRaisedDumb.usesSmartColor = NO;
    [bfRaisedDumb setBackgroundColor:[UIColor colorWithRed:33.f/255.f green:150.f/255.f blue:243.f/255.f alpha:1]];
    [bfRaisedDumb setTitle:@"BFPaperButton Raised: !Smart Color" forState:UIControlStateNormal];
    [bfRaisedDumb setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    [bfRaisedDumb addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:bfRaisedDumb];
    
    BFPaperButton *bfRaisedSmartSmall = [[BFPaperButton alloc] initWithFrame:CGRectMake(20, 375, 135, 83) raised:YES];
    [bfRaisedSmartSmall setBackgroundColor:[UIColor colorWithRed:33.f/255.f green:150.f/255.f blue:243.f/255.f alpha:1]];
    bfRaisedSmartSmall.titleLabel.numberOfLines = 0;
    bfRaisedSmartSmall.titleLabel.font = [UIFont systemFontOfSize:10.f];
    [bfRaisedSmartSmall setTitle:@"BFPaperButton Raised: Smart Color" forState:UIControlStateNormal];
    [bfRaisedSmartSmall setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    [bfRaisedSmartSmall addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:bfRaisedSmartSmall];
    
    BFPaperButton *bfRaisedDumbSmall = [[BFPaperButton alloc] initWithFrame:CGRectMake(163, 375, 135, 83) raised:YES];
    bfRaisedDumbSmall.usesSmartColor = NO;
    [bfRaisedDumbSmall setBackgroundColor:[UIColor colorWithRed:33.f/255.f green:150.f/255.f blue:243.f/255.f alpha:1]];
    bfRaisedDumbSmall.titleLabel.numberOfLines = 0;
    bfRaisedDumbSmall.titleLabel.font = [UIFont systemFontOfSize:10.f];
    bfRaisedDumbSmall.tapCircleDiameter = bfPaperButton_tapCircleDiameterFull;
    [bfRaisedDumbSmall setTitle:@"BFPaperButton Raised: !Smart Color, large circle" forState:UIControlStateNormal];
    [bfRaisedDumbSmall setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    [bfRaisedDumbSmall addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:bfRaisedDumbSmall];
    
    
    
    BFPaperButton *circle1 = [[BFPaperButton alloc] initWithFrame:CGRectMake(20, 468, 86, 86) raised:YES];
    [circle1 setBackgroundColor:[UIColor colorWithRed:33.f/255.f green:150.f/255.f blue:243.f/255.f alpha:1]];
    [circle1 setTitle:@"Center" forState:UIControlStateNormal];
    [circle1 setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    [circle1 setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [circle1 setTitleColor:[UIColor whiteColor] forState:UIControlStateHighlighted];
    [circle1 addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    circle1.cornerRadius = circle1.frame.size.width / 2;
    circle1.rippleFromTapLocation = NO;
    [self.view addSubview:circle1];
    
    BFPaperButton *circle2 = [[BFPaperButton alloc] initWithFrame:CGRectMake(116, 468, 86, 86) raised:YES];
    [circle2 setTitle:@"Custom" forState:UIControlStateNormal];
    [circle2 setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    [circle2 setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [circle2 setTitleColor:[UIColor whiteColor] forState:UIControlStateHighlighted];
    [circle2 addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    circle2.backgroundColor = [UIColor colorWithRed:0.3 green:0 blue:1 alpha:1];
    circle2.tapCircleColor = [UIColor colorWithRed:1 green:0 blue:1 alpha:0.6];  // Setting this color overrides "Smart Color".
    circle2.cornerRadius = circle2.frame.size.width / 2;
    circle2.rippleFromTapLocation = NO;
    circle2.rippleBeyondBounds = YES;
    circle2.tapCircleDiameter = MAX(circle2.frame.size.width, circle2.frame.size.height) * 1.3;
    [self.view addSubview:circle2];

    BFPaperButton *circle3 = [[BFPaperButton alloc] initWithFrame:CGRectMake(212, 468, 86, 86) raised:NO];
    [circle3 setTitle:@"Custom" forState:UIControlStateNormal];
    [circle3 setTitleFont:[UIFont fontWithName:@"HelveticaNeue-Light" size:15.f]];
    [circle3 setTitleColor:[UIColor colorWithRed:33.f/255.f green:33.f/255.f blue:33.f/255.f alpha:1] forState:UIControlStateNormal];
    [circle3 setTitleColor:[UIColor colorWithRed:33.f/255.f green:33.f/255.f blue:33.f/255.f alpha:1] forState:UIControlStateHighlighted];
    [circle3 addTarget:self action:@selector(buttonWasPressed:) forControlEvents:UIControlEventTouchUpInside];
    circle3.cornerRadius = circle3.frame.size.width / 2;
    circle3.tapCircleDiameter = 53;
    circle3.tapCircleColor = [UIColor colorWithRed:0.3 green:0 blue:1 alpha:0.6];  // Setting this color overrides "Smart Color".
    circle3.backgroundFadeColor = [UIColor colorWithRed:0 green:0 blue:1 alpha:0.3f]; // Setting this color overrides "Smart Color".
    [self.view addSubview:circle3];
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)buttonWasPressed:(id)sender
{
    NSLog(@"%@ was pressed!", ((UIButton *)sender).titleLabel.text);
}

@end
